### PR TITLE
[Clang] fix static operator()/[] call not evaluating object

### LINKED ISSUE #78355: [Clang] `E[...]`/`E(...)` does not evaluate `E` if it calls a `static` overload
Example:

```c++
struct X {
    X(int x) {
        __builtin_printf("X(%d)\n", x);
    }

    static void operator()() {
        __builtin_printf("operator()\n");
    }

    static void operator[]() {
        __builtin_printf("operator[]\n");
    }

    static void f(int) {
        __builtin_printf("f\n");
    }
};

int main() {
    X(__builtin_printf("main()1\n"))();
    X(__builtin_printf("main() 2\n"))[];
    X(__builtin_printf("main()  3\n")).f(0);
}
```

Prints:

```
operator()
operator[]
main()  3
X(10)
f
```

Instead of the expected:

```
main()1
X(8)
operator()
main() 2
X(9)
operator[]
main()  3
X(10)
f
```

https://godbolt.org/z/sv14fjMYK

### PR DESCRIPTION
**Pull request by MitalAshok**
Fixes #78355

### CODE REVIEW

File: clang/test/SemaCXX/overloaded-operator.cpp [hunk 1]
```diff
@@ -598,3 +600,27 @@ namespace B {
 }
 void g(B::X x) { A::f(x); }
 }
+
+namespace static_operator {
+#if __cplusplus >= 201703L
```
Comment (frederick-vs-ja) [inline clang/test/SemaCXX/overloaded-operator.cpp:605]: Why C++17?

**Review by frederick-vs-ja — COMMENTED** [review]
It seems that more things are being done #68485. Have you double checked that PR?

File: clang/test/SemaCXX/overloaded-operator.cpp (hunk 1 above)
Comment (MitalAshok) [reply clang/test/SemaCXX/overloaded-operator.cpp:605]: No way to modify things in a C++11/14 constexpr function (`++x` is not a constant expression), so no way to make a positive test that the expression was evaluated. Though I guess I should have done `(non_constexpr_fn(), X())()` and see if that failed.

**Comment by MitalAshok** [comment]
This pull request changes `E(...)` to produce the same AST as `E.operator()(...)`, which is similar to `E.f(...)` for a static member function `f`.

But the linked #68485 changes this to CXXOperatorCallExpr, which is more appropriate since it is used for non-member call operators like anyways

File: clang/test/SemaCXX/overloaded-operator.cpp (hunk 1 above)
Comment (frederick-vs-ja) [reply clang/test/SemaCXX/overloaded-operator.cpp:605]: > No way to modify things in a C++11/14 constexpr function (`++x` is not a constant expression), so no way to make a positive test that the expression was evaluated. Though I guess I should have done `(non_constexpr_fn(), X())()` and see if that failed.

Oh I see. It's inconvenient that runtime results can't be easily asserted in this test file.

I think the related changes of constexpr were made in C++14 ([N3652](https://wg21.link/n3652)).

**Comment by cor3ntin** [comment]
@MitalAshok Can we close this?

**Pull request closed without merging**